### PR TITLE
GODRIVER-41 implement sdam monitoring

### PR DESCRIPTION
**Pull request by iwysiu**
Co-authored-by: Alice Thum <alice.thum@mongodb.com>

### CODE REVIEW

File: event/monitoring.go [hunk 1]
```diff
@@ -89,3 +92,75 @@ type PoolEvent struct {
...
+
+// ServerHeartbeatFailedEvent is an event generated when the ismaster fails.
+type ServerHeartbeatFailedEvent struct {
+	Duration     int64
+	Reply        error
+	ConnectionID string
+	Awaited      bool
+}
+
+// SdamMonitor represents a monitor that is triggered for different SDAM events.
```
Comment (divjotarora) [inline event/monitoring.go:155]: This type of monitoring is pretty complex and requires understanding all of the stuff that goes on in the driver's background routines to write effective handlers for the events. Can you add some more high-level documentation here or in `ClientOptions` and file a follow-on ticket to add a `event/doc.go` file that explains all of the different types of monitoring in more detail?

File: x/mongo/driver/topology/sdam_spec_test.go [hunk 2]
```diff
@@ -404,6 +494,109 @@ func applyErrors(t *testing.T, topo *Topology, errors []applicationError) {
 	}
 }
 
+func compareServerDescriptions(t *testing.T,
```
Comment (divjotarora) [inline x/mongo/driver/topology/sdam_spec_test.go:497]: nit: Use `expected` and `actual` rather than `e` and `a` for clarity.

File: x/mongo/driver/topology/server.go [hunk 3]
```diff
@@ -653,12 +657,14 @@ func (s *Server) check() (description.Server, error) {
 		heartbeatConn := initConnection{s.conn}
 		baseOperation := s.createBaseOperation(heartbeatConn)
 		previousDescription := s.Description()
+		streamable := previousDescription.TopologyVersion != nil
 
 		switch {
 		case s.conn.getCurrentlyStreaming():
+
```
Comment (divjotarora) [inline x/mongo/driver/topology/server.go:664]: The `duration` variable needs to be set in this block. Per the SDAM monitoring spec, when streaming responses, the duration should account for the time spent waiting for the server response. This means that the duration isn't an accurate representation of server RTT in this case. I think we can simplify a bit by avoid the duplicated time calculations:

```
if descPtr == nil && err == nil {
   // using existing connection
   
   // setup stuff...
   
   start = time.Now()
   switch {
       case s.conn.getCurrentlyStreaming():
       case streamable:
       default:
   }
   duration = time.Since(start)
}
```

None of these cases do any blocking work besides the isMaster execution, so it should be fine to measure the time this way.

File: x/mongo/driver/topology/server.go [hunk 4]
```diff
@@ -781,6 +803,85 @@ func (ss *ServerSubscription) Unsubscribe() error {
 	return nil
 }
 
+// publishes a ServerDescriptionChangedEvent to indicate the server description has changed
+func (s *Server) publishServerDescriptionChangedEvent(prev *description.Server, current *description.Server) {
```
Comment (divjotarora) [inline x/mongo/driver/topology/server.go:807]: Can all of the event-publishing helpers that require server descriptions take a value rather than a pointer? The code that does the actual monitoring check keeps the descriptions as pointers because it needs to tell if it's been set, but that's an implementation detail that I'd like to avoid leaking to other helpers. Once we get to these helpers, we know the description has been set and can use it without taking a pointer.

**Review by divjotarora — COMMENTED** [review]
I haven't reviewed all of the tests yet. Mostly small things, but there are some bugs with the event publishing when we're doing the streamable protocol.

File: event/monitoring.go (hunk 1 above)
Comment (iwysiu) [reply event/monitoring.go]: this could probably be improved. created GODRIVER-1747

File: x/mongo/driver/topology/topology.go [hunk 5]
```diff
@@ -160,9 +151,34 @@ func (t *Topology) Connect() error {
...
+	}
+
+	// store new description
+	newDesc := description.Topology{
+		Kind:                  t.fsm.Kind,
+		Servers:               t.fsm.Servers,
+		SessionTimeoutMinutes: t.fsm.SessionTimeoutMinutes,
+	}
+	t.desc.Store(newDesc)
+	t.publishTopologyDescriptionChangedEvent(description.Topology{}, t.fsm.Topology)
```
Comment (kevinAlbs) [inline x/mongo/driver/topology/topology.go:179]: Since this occurs while `serversLock` is held, I assume if a user callback did anything requiring server selection on the same client. This is inconsistent behavior among drivers. Callbacks for libmongoc similarly would deadlock. If that is the case, I suggest noting in the docs for the callbacks that the callbacks should not attempt to do any operation on the same client requiring server selection. See also: https://jira.mongodb.org/browse/DRIVERS-1277

**Review by kevinAlbs — CHANGES_REQUESTED** [review]
This is exciting! I have some questions/comments about the interface and server description equality. I will review tests on my next pass.

Note, I created GODRIVER-1749 to do in follow-up for SDAM integration tests assertions. It seems out of scope of this ticket/PR.

File: event/monitoring.go [hunk 6]
```diff
@@ -89,3 +92,78 @@ type PoolEvent struct {
...
+type TopologyOpeningEvent struct {
+	TopologyID primitive.ObjectID
+}
+
+// TopologyClosedEvent is an event generated when the topology is closed.
+type TopologyClosedEvent struct {
+	TopologyID primitive.ObjectID
+}
+
+// ServerHeartbeatStartedEvent is an event generated when the ismaster command is started.
```
[the remark this answers is not in the document]
Comment (iwysiu) [reply event/monitoring.go:133]: done

File: event/monitoring.go [hunk 7]
```diff
@@ -89,3 +92,78 @@ type PoolEvent struct {
...
+	Duration     int64
+	Reply        description.Server
+	ConnectionID string
+	Awaited      bool
+}
+
+// ServerHeartbeatFailedEvent is an event generated when the ismaster fails.
+type ServerHeartbeatFailedEvent struct {
+	Duration     int64
+	Reply        error
```
[the remark this answers is not in the document]
Comment (iwysiu) [reply event/monitoring.go:150]: done

File: event/monitoring.go [hunk 8]
```diff
@@ -89,3 +92,78 @@ type PoolEvent struct {
 type PoolMonitor struct {
 	Event func(*PoolEvent)
 }
+
+// ServerDescriptionChangedEvent represents a server description change.
+type ServerDescriptionChangedEvent struct {
+	Address             address.Address
```
[the remark this answers is not in the document]
Comment (iwysiu) [reply event/monitoring.go:98]: from what I've seen, it is used, but only for things that aren't obvious. If you think any of the ones I've skipped need explanations I can add it.

File: vendor/github.com/aws/aws-sdk-go/internal/ini/ini_parser.go [hunk 9]
```diff
@@ -37,23 +37,23 @@ const (
 
 // parseTable is a state machine to dictate the grammar above.
 var parseTable = map[ASTKind]map[TokenType]int{
-	ASTKindStart: map[TokenType]int{
```
[the remark this answers is not in the document]
Comment (iwysiu) [reply vendor/github.com/aws/aws-sdk-go/internal/ini/ini_parser.go:40]: no, I assume that was gofmt

File: x/mongo/driver/topology/server.go [hunk 10]
```diff
@@ -676,23 +685,29 @@ func (s *Server) check() (description.Server, error) {
...
 			s.conn.setSocketTimeout(s.cfg.heartbeatTimeout)
+
 			err = baseOperation.Execute(s.heartbeatCtx)
 		}
+		duration = int64(time.Since(start))
+
 		if err == nil {
 			tempDesc := baseOperation.Result(s.address)
 			descPtr = &tempDesc
+			s.publishServerHeartbeatSucceededEvent(s.conn.ID(), duration, tempDesc, streamable)
```
Comment (divjotarora) [inline x/mongo/driver/topology/server.go:703]: When calling `publishServerHeartbeatStartedEvent`, you use `s.conn.getCurrentlyStreaming() || streamable` as the value for the `await` field but here you just use `streamable`. Are these intentionally different?

File: vendor/github.com/aws/aws-sdk-go/internal/ini/ini_parser.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (divjotarora) [reply vendor/github.com/aws/aws-sdk-go/internal/ini/ini_parser.go:40]: gofmt wouldn't be responsible for something like this. If these changes weren't intentional, let's remove them.

File: event/monitoring.go [hunk 11]
```diff
@@ -96,60 +96,60 @@ type PoolMonitor struct {
...
+// ServerHeartbeatStartedEvent is an event generated when the heartbeat is started.
 type ServerHeartbeatStartedEvent struct {
-	ConnectionID string
-	Awaited      bool
+	ConnectionID string // The address this heartbeat was sent to with a unique identifier
+	Awaited      bool // If this heartbeat was awaitable
 }
 
-// ServerHeartbeatSucceededEvent is an event generated when the ismaster succeeds.
+// ServerHeartbeatSucceededEvent is an event generated when the hearbeat succeeds.
```
Comment (divjotarora) [inline event/monitoring.go:139]: hearbeat -> heartbeat

File: mongo/description/server.go [hunk 12]
```diff
@@ -425,7 +425,7 @@ func (s Server) SupportsRetryWrites() bool {
 
 // Equal compares two server descriptions and returns true if they are equal
 func (s Server) Equal(other Server) bool {
-	if s.Addr.String() != other.Addr.String() {
+	if s.CanonicalAddr.String() != other.CanonicalAddr.String() {
```
Comment (divjotarora) [inline mongo/description/server.go:428]: Why this change?

File: mongo/description/server.go [hunk 13]
```diff
@@ -459,7 +459,34 @@ func (s Server) Equal(other Server) bool {
 		return false
 	}
 
-	// the rest of the server fields are not used in sdam monitoring server descriptions and not compared
+	if s.LastError != other.LastError {
```
Comment (divjotarora) [inline mongo/description/server.go:462]: This isn't a safe way to compare errors. Errors generated via mechanisms like `errors.New` won't be equal even if they have the same message. This can come up in unexpected ways because some of these errors come from external sources like Go's `net` package. See https://play.golang.org/p/jwAe6j1mH1l for an example.

File: mongo/description/version_range.go [hunk 14]
```diff
@@ -25,6 +25,17 @@ func (vr VersionRange) Includes(v int32) bool {
...
 
+// Equals returns a bool indicating whether the supplied VersionRange is equal.
+func (vr *VersionRange) Equals(other *VersionRange) bool {
+	if vr == nil && other == nil {
+		return true
+	}
+	if vr == nil || other == nil {
+		return false
+	}
+	return vr.Min == other.Min && vr.Max <= other.Max
```
Comment (divjotarora) [inline mongo/description/version_range.go:36]: Why is the check `vr.Max <= other.Max`? If the server's `maxWireVersion` changes, it means the server has either upgraded or downgraded. That definitely seems like a situation where the new description would not be equal to the old one.

File: x/mongo/driver/topology/server_test.go [hunk 15]
```diff
@@ -309,6 +312,90 @@ func TestServer(t *testing.T) {
...
+			publishedEvents = append(publishedEvents, *e)
+		}
+
+		sdam := &event.ServerMonitor{
+			ServerHeartbeatStarted:   serverHeartbeatStarted,
+			ServerHeartbeatSucceeded: serverHeartbeatSucceeded,
+			ServerHeartbeatFailed:    serverHeartbeatFailed,
+		}
+
+		// test that client metadata is sent on handshakes but not heartbeats
```
Comment (divjotarora) [inline x/mongo/driver/topology/server_test.go:336]: I think this comment is copy/pasted from another test?

Comment (iwysiu) [reply x/mongo/driver/topology/server_test.go]: done

File: x/mongo/driver/topology/server_test.go [hunk 16]
```diff
@@ -309,6 +312,89 @@ func TestServer(t *testing.T) {
...
+				t.Fatalf("error adding response: %v", err)
+			}
+			_, err = s.check()
+			assert.Nil(t, err, "check error: %v", err)
+
+			assert.Equal(t, len(publishedEvents), 2, "expected %v events, got %v", 2, len(publishedEvents))
+
+			started, ok := publishedEvents[0].(event.ServerHeartbeatStartedEvent)
+			assert.True(t, ok, "expected type %T, got %T", event.ServerHeartbeatStartedEvent{}, publishedEvents[0])
+			assert.Equal(t, started.ConnectionID, s.conn.ID(), "expected connectionID to match")
```
Comment (kevinAlbs) [inline x/mongo/driver/topology/server_test.go:373]: Since `makeIsMasterReply` returns `{'ok': 1}` I suggest also asserting `started.Awaited` is false.

File: x/mongo/driver/topology/server_test.go [hunk 17]
```diff
@@ -309,6 +312,89 @@ func TestServer(t *testing.T) {
...
+
+			succeeded, ok := publishedEvents[1].(event.ServerHeartbeatSucceededEvent)
+			assert.True(t, ok, "expected type %T, got %T", event.ServerHeartbeatSucceededEvent{}, publishedEvents[1])
+			assert.Equal(t, succeeded.ConnectionID, s.conn.ID(), "expected connectionID to match")
+			assert.Equal(t, succeeded.Reply.Addr, s.address, "expected address %T, got %T", s.address, succeeded.Reply.Addr)
```
Comment (kevinAlbs) [inline x/mongo/driver/topology/server_test.go:378]: Assert `succeeded.Awaited` is false.

File: x/mongo/driver/topology/server_test.go [hunk 18]
```diff
@@ -309,6 +312,89 @@ func TestServer(t *testing.T) {
...
+			// do a heartbeat with a non-nil connection
+			channelConn.ReadErr <- errors.New("error")
+			_, err = s.check()
+			assert.Nil(t, err, "check error: %v", err)
+
+			assert.Equal(t, len(publishedEvents), 2, "expected %v events, got %v", 2, len(publishedEvents))
+
+			started, ok := publishedEvents[0].(event.ServerHeartbeatStartedEvent)
+			assert.True(t, ok, "expected type %T, got %T", event.ServerHeartbeatStartedEvent{}, publishedEvents[0])
+			assert.Equal(t, started.ConnectionID, s.conn.ID(), "expected connectionID to match")
```
Comment (kevinAlbs) [inline x/mongo/driver/topology/server_test.go:391]: Assert `started.Awaited` is false.

File: x/mongo/driver/topology/server_test.go [hunk 19]
```diff
@@ -309,6 +312,89 @@ func TestServer(t *testing.T) {
...
+
+			failed, ok := publishedEvents[1].(event.ServerHeartbeatFailedEvent)
+			assert.True(t, ok, "expected type %T, got %T", event.ServerHeartbeatFailedEvent{}, publishedEvents[1])
+			assert.Equal(t, failed.ConnectionID, s.conn.ID(), "expected connectionID to match")
```
Comment (kevinAlbs) [inline x/mongo/driver/topology/server_test.go:395]: Assert `failed.Awaited` is false and that `failed.Failure` is equal to the expected error (if that's reasonable).

**Review by kevinAlbs — APPROVED** [review]
LGTM with some minor test additions.

File: mongo/description/server_test.go [hunk 20]
```diff
@@ -0,0 +1,70 @@
+// Copyright (C) MongoDB, Inc. 2017-present.
```
[the remark this answers is not in the document]
Comment (divjotarora) [reply mongo/description/server_test.go]: I wouldn't do this. These licenses are added via a script and we want to keep them the same across files. Also, why does this need to be changed? The license says the copyright is starting in 2017 and continues to present day.

File: x/mongo/driver/topology/server_test.go [hunk 21]
```diff
@@ -309,6 +312,89 @@ func TestServer(t *testing.T) {
...
+
+		dialer := &channelNetConnDialer{}
+		dialerOpt := WithDialer(func(Dialer) Dialer {
+			return dialer
+		})
+		serverOpts := []ServerOption{
+			WithConnectionOptions(func(connOpts ...ConnectionOption) []ConnectionOption {
+				return append(connOpts, dialerOpt)
+			}),
+			withMonitoringDisabled(func(bool) bool { return true }),
```
[the remark this answers is not in the document]
Comment (divjotarora) [reply x/mongo/driver/topology/server_test.go:343]: @kevinAlbs I think there's some confusion here. This disables both the server monitoring and the RTT monitor. This test manually simulates the server monitor by calling `s.check()` below.

File: mongo/description/server_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply mongo/description/server_test.go]: Gotcha, I had assumed the starting year was when the file was added.

File: x/mongo/driver/topology/server_test.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply x/mongo/driver/topology/server_test.go:343]: Gotcha, thanks for clarifying.

**Comment by divjotarora** [comment]
@iwysiu Can we change the license back to 2017?

File: mongo/description/server.go [hunk 22]
```diff
@@ -412,3 +422,87 @@ func decodeStringMap(element bson.RawElement, name string) (map[string]string, e
...
 	return s.SessionTimeoutMinutes != 0 && s.Kind != Standalone
 }
+
+// Equal compares two server descriptions and returns true if they are equal
+func (s Server) Equal(other Server) bool {
+	if s.CanonicalAddr.String() != other.CanonicalAddr.String() {
+		return false
+	}
+
+	if len(s.Arbiters) != len(other.Arbiters) ||
```
Comment (divjotarora) [inline mongo/description/server.go:432]: These `len` checks aren't needed anymore because `stringSliceEquals` will handle them.

File: x/mongo/driver/topology/server_test.go [hunk 23]
```diff
@@ -309,6 +312,94 @@ func TestServer(t *testing.T) {
...
+		channelConn := s.conn.nc.(*drivertest.ChannelNetConn)
+		_ = channelConn.GetWrittenMessage()
+
+		t.Run("success", func(t *testing.T) {
+			publishedEvents = nil
+			// do a heartbeat with a non-nil connection
+			if err = channelConn.AddResponse(makeIsMasterReply()); err != nil {
+				t.Fatalf("error adding response: %v", err)
+			}
+			_, err = s.check()
```
Comment (divjotarora) [inline x/mongo/driver/topology/server_test.go:366]: I described this issue in Slack a bit, but I'll summarize here so it's recorded in the PR. All calls to `check` have to then call `GetWrittenMessage` to clear the connection's write buffer. If this isn't done, the next `failure` test will fail to write anything to the buffer because it's full.

File: x/mongo/driver/topology/server_test.go [hunk 24]
```diff
@@ -309,6 +312,94 @@ func TestServer(t *testing.T) {
...
+			succeeded, ok := publishedEvents[1].(event.ServerHeartbeatSucceededEvent)
+			assert.True(t, ok, "expected type %T, got %T", event.ServerHeartbeatSucceededEvent{}, publishedEvents[1])
+			assert.Equal(t, succeeded.ConnectionID, s.conn.ID(), "expected connectionID to match")
+			assert.Equal(t, succeeded.Reply.Addr, s.address, "expected address %v, got %v", s.address, succeeded.Reply.Addr)
+			assert.False(t, succeeded.Awaited, "expected awaited to be false")
+		})
+		t.Run("failure", func(t *testing.T) {
+			publishedEvents = nil
+			// do a heartbeat with a non-nil connection
+			channelConn.ReadErr <- errors.New("error")
```
Comment (divjotarora) [inline x/mongo/driver/topology/server_test.go:385]: I think we can make this more robust by doing:

```
readErr := errors.New("error")
channelConn.ReadErr <- readErr

...

assert.True(t, errors.Is(failed.Failure, readErr), ...)
```

The use of `errors.Is` will require build-flagging this test file on only run on Go 1.13+.

File: x/mongo/driver/topology/server.go [hunk 25]
```diff
@@ -682,17 +690,22 @@ func (s *Server) check() (description.Server, error) {
 			// execute a regular heartbeat without any additional parameters.
 
 			s.conn.setSocketTimeout(s.cfg.heartbeatTimeout)
+
```
Comment (divjotarora) [inline x/mongo/driver/topology/server.go:693]: nit: This blank line wasn't removed.

File: mongo/description/server.go (hunk 22 above)
Comment (iwysiu) [reply mongo/description/server.go]: done

File: x/mongo/driver/topology/server.go (hunk 25 above)
Comment (iwysiu) [reply x/mongo/driver/topology/server.go]: done

File: x/mongo/driver/topology/server_test.go (hunk 23 above)
Comment (iwysiu) [reply x/mongo/driver/topology/server_test.go:368]: done

File: x/mongo/driver/topology/server_test.go (hunk 24 above)
Comment (iwysiu) [reply x/mongo/driver/topology/server_test.go]: done

**Review by divjotarora — APPROVED** [review]
